File: system-design/LRU-cache/LRU-cache.ts
/**
 * 
Design and implement a data structure for Least Recently Used (LRU) cache.

It should support the following operations: get and set.

- get(key)
  Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.

- set(key, value)
  Set or insert the value if the key is not already present.
  When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.
  Finally, you need to return the data from each get.

// Assume key is value;
 */

// https://www.jiuzhang.com/solution/lru-cache/

export class LRUCache {
  private map: Record<string, Node>;
  private head: Node; // Maybe it is better to setup dummy node.
  private tail: Node;
  private capacity = 0;

  constructor(capacity: number) {
    this.capacity = capacity;
  }

  get(key: string) {
    if (this.map[key]) {
      this.shiftToTail(this.map[key]);
      return this.map[key].value;
    } else {
      return null;
    }
  }

  set(key: string, value: any) {
    // if key exists
    if (this.map[key]) {
      // shift
      this.shiftToTail(this.map[key]);
    }
    // if key does not exist
    else {
      const node = new Node(key, value);
      this.map[key] = node;

      if (this.head === null) {
        this.head === node;
        this.tail === node;
      } else {
        this.shiftToTail(node);
      }

      if (Object.keys(this.map).length > this.capacity) {
        const headToRemove = this.head;

        this.head = this.head.next;
        this.head.previous = null;

        delete this.map[headToRemove.key];
      }
    }
  }

  private shiftToTail(node: Node) {
    if (node.previous) {
      node.previous.next = node.next;
    }

    if (node.next) {
      node.next.previous = node.previous;
    }

    // add node to end;
    node.previous = this.tail;
    node.next = null;
    this.tail.next = node;
    this.tail = node;
  }
}

export class Node {
  previous: Node;
  next: Node;
  key: any;
  value: any;

  constructor(key: any, value: any) {
    this.key = key;
    this.value = value;
    this.next = null;
    this.previous = null;
  }
}

File: system-design/system-design/LRU-cache/LRUCache.java
import java.util.HashMap;

/**
 * implement following method for an LRU cache;
 *
 * Company Amazon
 */
public class LRUCache {
    private HashMap<Integer, Node> map;
    private int cap;
    private int count;
    private Node head;
    private Node tail;

    public static void main(String[] args) {
        LRUCache lru = new LRUCache(5);

        lru.put(1, 1);
        lru.put(2, 2);
        lru.put(3, 3);
        lru.put(4, 4);
        lru.put(5, 5);
        lru.put(6, 6);

        // expect 1 to be deleted
        System.out.println(lru.get(1));

        lru.put(3, 33);

        System.out.println(lru.head.next.val);
    }

    public LRUCache(int capacity) {
        map = new HashMap<>();

        cap = capacity;
        count = 0;

        head = new Node(-1, -1);
        tail = new Node(-1, -1);

        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        if (map.containsKey(key)) {
            Node node = map.get(key);

            delete(node);
            addToHead(node);

            return node.val;
        }
        return -1;
    }

    public void put(int key, int value) {
        // if key exists, we update value, move to the front
        if (map.containsKey(key)) {
            Node node = map.get(key);

            node.val = value;

            delete(node);
            addToHead(node);
        }

        // if key does not exist, create a node, add to map, add to the front, update capcity
        else {
            Node node = new Node(key, value);
            map.put(key, node);
            addToHead(node);

            if(count == cap) {
                map.remove(tail.prev.key);
                delete(tail.prev);
                count--;
            }
            count++;
        }
    }

    public void delete(Node node) {
        //System.out.println(node.val);
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    public void addToHead(Node node) {
        node.next = head.next;
        head.next = node;
        node.prev = head;
        node.next.prev = node;
    }

    private class Node {
        int key;
        int val;
        Node next;
        Node prev;

        public Node(int key, int value) {
            this.key = key;
            val = value;
            next = null;
            prev = null;
        }
    }
}
